### PR TITLE
Add email-sending support as SMTP client

### PR DESCRIPTION
**Pull request by fantix**
- [x] SMTP server config
- [x] Email sending API
- [x] Manual testing
- [x] Add unit tests
- [x] Integrate with auth for sending password reset email

### CODE REVIEW

File: edb/server/ext/smtp.py [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+import asyncio
+import email
+
+import aiosmtplib
+
+from edb.common import retryloop
+from edb.ir import statypes
+
+from . import util
+from ... import errors
```
Comment (1st1) [inline edb/server/ext/smtp.py:30]: Please use absolute imports for when we're importing from parents and grandparents

File: edb/lib/ext/smtp.edgeql [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        create required property security -> ext::smtp::ConnectionSecurity {
+            set default := ext::smtp::ConnectionSecurity.PreferSTARTTLS;
+        };
+        create required property validate_certs -> std::bool {
+            set default := true;
+        };
+        create required property concurrency -> std::int32 {
+            create constraint std::min_value(1);
+            set default := 2;
+        };
```
Comment (1st1) [inline edb/lib/ext/smtp.edgeql:41]: I saw how you use it in code, but I'm not sure I fully understand it. How does configuring `concurrency` affects things? Can we hide this config option and auto-configure it on the server somehow?

Comment (fantix) [reply edb/lib/ext/smtp.edgeql:41]: This limits the number of maximum concurrent SMTP connections - for example, setting this value to `1` will serialize all SMTP requests. So this is basically a customer feature to be able to play nice with their SMTP providers (ideally, there should also be a rate limit).

Comment (fantix) [reply edb/lib/ext/smtp.edgeql:41]: The caller will block until the email is sent.

File: edb/server/protocol/auth_ext/ui.py [hunk 3]
```diff
@@ -409,3 +411,43 @@ def _render_button(text: str):
...
+    )
+    alternative.attach(plain_text_msg)
+    html_msg = mime_text.MIMEText(
+        f"""
+<!DOCTYPE html>
+<html>
+  <body>
+    <a href="{reset_url}">Reset password</a>
+  </body>
+</html>
```
Comment (fantix) [inline edb/server/protocol/auth_ext/ui.py:446]: We need proper title / plain text / HTML email content here.

File: tests/test_http_ext_auth.py [hunk 4]
```diff
@@ -315,6 +317,21 @@ class TestHttpExtAuth(tb.ExtAuthTestCase):
...
+        ext::auth::SMTPConfig::timeout_per_attempt := <duration>'1 seconds';
+
+        CONFIGURE CURRENT DATABASE SET
+        ext::auth::SMTPConfig::timeout_per_email := <duration>'1 seconds';
+
+        CONFIGURE CURRENT DATABASE SET
+        ext::auth::SMTPConfig::validate_certs := true;
+
+        CONFIGURE CURRENT DATABASE SET
+        ext::auth::SMTPConfig::security := ext::auth::SMTPSecurity.TLS;
```
Comment (fantix) [inline tests/test_http_ext_auth.py:333]: I was expecting the default values to kick in without setting these manually, but it turned out not the case.

Comment (elprans) [reply tests/test_http_ext_auth.py:333]: Hm, this is odd. @msullivan, any idea why?

File: edb/server/protocol/auth_ext/smtp.py [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+    password = util.maybe_get_config(
+        db.db_config,  # type: ignore
+        "ext::auth::SMTPConfig::password",
+    )
+    timeout_per_attempt = util.get_config(
+        db.db_config,  # type: ignore
+        "ext::auth::SMTPConfig::timeout_per_attempt",
+        expected_type=statypes.Duration,
+    )
+    req_timeout = timeout_per_attempt.to_microseconds() / 1_000_000.0
```
Comment (scotttrinh) [inline edb/server/protocol/auth_ext/smtp.py:77]: `req_timeout` is measured in millionths of a microsecond?

**Review by scotttrinh — COMMENTED** [review]
Looks really good so far! One clarifying question.

File: edb/server/protocol/auth_ext/smtp.py (hunk 5 above)
Comment (fantix) [reply edb/server/protocol/auth_ext/smtp.py:77]: Yes I believe so: https://www.postgresql.org/docs/current/datatype-datetime.html#DATATYPE-INTERVAL-INPUT

Comment (fantix) [reply edb/server/protocol/auth_ext/smtp.py:77]: Ah, should use `*` - good catch!

**Review by scotttrinh — APPROVED** [review]

File: edb/lib/ext/auth.edgeql [hunk 6]
```diff
@@ -144,6 +144,59 @@ CREATE EXTENSION PACKAGE auth VERSION '1.0' {
                 indicates that the token should never expire.";
             set default := <std::duration>'336 hours';
         };
+
+    };
+
+    create scalar type ext::auth::SMTPSecurity extending enum<PlainText, TLS, STARTTLS, STARTTLSOrPlainText>;
```
Comment (1st1) [inline edb/lib/ext/auth.edgeql:150]: Should it be `StartTLS`?

Comment (fantix) [reply edb/lib/ext/auth.edgeql:150]: Yeah I'm seeing both in different places. [RFC 3207](https://www.ietf.org/rfc/rfc3207.txt) uses "STARTTLS".

File: edb/server/protocol/auth_ext/smtp.py [hunk 7]
```diff
@@ -0,0 +1,152 @@
...
+        email.message.Message,
+        str,
+        bytes,
+    ],
+    sender: Optional[str] = None,
+    recipients: Optional[Union[str, Sequence[str]]] = None,
+    test_mode: bool = False,
+) -> None:
+    global _semaphore
+    if _semaphore is None:
```
Comment (1st1) [inline edb/server/protocol/auth_ext/smtp.py:50]: So this is a global semaphore for all tenants?

Comment (fantix) [reply edb/server/protocol/auth_ext/smtp.py:50]: Yes that's correct.

File: tests/test_http_ext_auth.py (hunk 4 above)
Comment (fantix) [reply tests/test_http_ext_auth.py:333]: This is probably because we didn't use the `spec.Setting.default` value in `auth/utils.py`, I'll push a fix.

**Pull request merged**